src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,652 @@
+     * particular {@code KDF} instance.
+     *
+     * @param derivationParameterSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return a byte array corresponding to a key built from the
+     *     KDF output and according to the derivation parameters. Rather than
+     *     permitting a {@code null} return value here, implementations will
+     *     handle cases where a value cannot be returned by throwing an
+     *     appropriate {@code Exception}.
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid
+     * @throws UnsupportedOperationException
+     *     if the derived key material is not extractable
+     * @throws NullPointerException
+     *     if {@code derivationParameterSpec} is null
+     */
+    public byte[] deriveData(AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+
+        Objects.requireNonNull(derivationParameterSpec);
+        if (spi != null) {
+            return spi.engineDeriveData(derivationParameterSpec);
+        } else {
+            return (byte[]) chooseProvider(null, derivationParameterSpec);
+        }
+    }
+
+    /**
+     * Selects the first provider in the list that supports the
+     * {@code KDFParameters}.
+     */
+    private void chooseFirstProvider() {
+        if ((spi != null) || (serviceIterator == null)) {
+            return;
+        }
+        synchronized (lock) {
+            if (spi != null) {
+                return;
+            }
+            Exception lastException = null;
+            while ((firstService != null) || serviceIterator.hasNext()) {
+                Service s;
+                if (firstService != null) {
+                    s = firstService;
+                    firstService = null;
+                } else {
+                    s = serviceIterator.next();
+                }
+                if (!JceSecurity.canUseProvider(s.getProvider())) {
+                    continue;
+                }
+                try {
+                    Object obj = s.newInstance(kdfParameters);
+                    if (!(obj instanceof KDFSpi)) {
+                        continue;
+                    }
+                    spi = (KDFSpi) obj;
+                    provider = s.getProvider();
+                    // not needed any more
+                    firstService = null;
+                    serviceIterator = null;
+                    return;
+                } catch (NoSuchAlgorithmException e) {
+                    lastException = e;
+                }
+            }
+            ProviderException e = new ProviderException(
+                "Could not construct KDFSpi instance");
+            if (lastException != null) {
+                e.initCause(lastException);
+            }
+            throw e;
+        }
+    }
```

Comment:
As discussed separately, if we enhance the logic of finding the `firstService` with the instantiation and instaneof KDFSpi check (on line 564-567), then we don't need to repeat the code here. We can just simply assign it to `spi`, set `serviceIterator` to `null` and be done with it.